src/box/iproto.cc

```diff
@@ -1857,7 +2069,46 @@ net_send_msg(struct cmsg *m)
 {
 	struct iproto_msg *msg = (struct iproto_msg *) m;
 	struct iproto_connection *con = msg->connection;
+	struct iproto_stream *stream = msg->stream;
+	struct iproto_msg *tmp;
+
+	if (stream == NULL)
+		goto send_msg;
+
+	tmp = stailq_shift_entry(&stream->pending_requests,
+				 struct iproto_msg, in_stream);
+	assert(tmp == msg);
+	(void)tmp;
+	errinj_stream_msg_count_add(-1);
+
+	if (stailq_empty(&stream->pending_requests)) {
+		struct mh_i64ptr_node_t node = { stream->id, NULL };
+		mh_i64ptr_remove(con->streams, &node, 0);
+		iproto_stream_delete(stream);
+	} else {
+		struct cmsg *m;
+		if (!evio_has_fd(&con->input)) {
+			m = &stream->on_disconnect;
+			struct iproto_thread *td = con->iproto_thread;
+			cmsg_init(m, td->destroy_stream_on_disconnect_route);
+		} else {
+			/*
+			 * If there are new messages for this stream
+			 * then schedule their processing.
+			 */
+			struct iproto_msg *next =
+				stailq_first_entry(&stream->pending_requests,
+						   struct iproto_msg,
+						   in_stream);
+			assert(next != NULL);
+			next->wpos = con->wpos;
+			m = &next->base;
+		}
+		cpipe_push_input(&con->iproto_thread->tx_pipe, m);
```

Comment:
Does this mean that the next queued message will be pushed only after the previous one has been processed, i.e. there's always an iproto-tx round trip between processing streamlined messages? If so, this means we have a serious performance issue by design. Are there any plans to fix it?

Anyway, why did you decide to maintain the stream hash table on the iproto side? IMO iproto should take care only of networking while streams is about request ordering, which can and should be handled by tx. What are the benefits of having the hash on the iproto side?

Reply:
Streams must process messages strictly sequentially, that is, the processing of the next message should not begin until the previous one has been processed. The performance problem that you are talking about can occur if we are talking about a transaction in a thread, and the performance loss will only be from the frequent capture of cbus mutex, since the processing in tx should be be strictly sequential. @Gerold103 has idea how to send send several messages to tx thread at once, during our discussion we decide to implement it in other patch.Stream hash was implemented on the iproto side because I try to do the least amount of work on the tx side.

Reply:
> Stream hash was implemented on the iproto side because I try to do the least amount of work on the tx side.

Lookup in a hash is virtually free in comparison to the amount of other work we do in tx so I disagree that this is a good reason.

> @Gerold103 has idea how to send send several messages to tx thread at once, during our discussion we decide to implement it in other patch.

I don't think we should commit this patch until we see the implementation of this optimization or at least a design document. Rationale: if the idea doesn't work out (e.g. out of sheer complexity), we end up with a wrongly designed feature, and it will take us longer to rework it then than now.

Or we should all agree that this performance issue is not critical and can be ignored.

Reply:
What are the worthy reasons to keep the streams solely in the tx thread?

By storing them in iproto thread we:
- Save tx thread time for not allocating/deleting streams, reallocating the stream hash table. Given how streams are created and deleted on demand, it might be not as free as you think. Streams might live just for begin + a couple of DML requests + commit, and each such combination will create/delete a stream. This includes not only the memory management I mentioned above, but also creation of the stream struct. Given how we see visible loss of perf when we add a few `if`s on a hot path in txn code, I wouldn't say it is all free. It adds work to practically each transaction in case Tarantool will be used 'without Lua'. As a transactional key-value storage. At least, this is the main usage for this feature AFAIU.
- We re-use the balancing of requests over the fiber pool. If we would have the queue of requests entirely in TX thread, you would need to handle the entire queue it in a single `cmsg_deliver` in `fiber_pool_f`. This might lead to starvation when some highly 'popular' streams take much more fiber time than other streams and independent messages. They simply won't let their fiber go for too long. Even if they yield. Unless we hack the fiber pool somehow to be able to 'reschedule' a request for later to do other requests handling and will limit the max number of stream messages to handle at once.

On the other hand if we would store the streams in TX thread, it probably would be simpler to handle disconnect (would be processed inside the existing `tx_process_disconnect` function). And it might be simpler to process the stream messages in batches in the future, so as not to push them into TX one by one.

Overall, the iproto-owned streams seem better to me so far.

Reply:
> What are the worthy reasons to keep the streams solely in the tx thread?

Discussed f2f. The main reason to handle streams in tx rather than in iproto is that stream id isn't really a property intrinsic to the protocol - it defines how requests are scheduled (executed) by tx (all requests with the same stream id should be scheduled sequentially and ideally by the same fiber). A good example by @alyapunov: if we add some other protocol (e.g. Redis), the streaming logic will work out-of-the-box if it is implemented on the tx side. If it is a part of the iproto, however, we will need to factor it out into a separate module somehow or simply rewrite it from the scratch, which is not good.

There may, however, be some problems regarding scheduling stream-lined requests in tx (e.g. starvation). @EvgenyMekhanik will look into this and try to come up with a scheduling algorithm which isn't affected by these issues, after which we will decide how to proceed.

Regarding potential performance problems, I don't think that a lookup in a stream hash on the tx side is going to be noticeable, because we do a lot of lookups (in space cache, in Lua tables) every time we execute a request. Besides, a potential performance degradation may only occur if the new feature is enabled. Ideally, the performance impact needs to be estimated before we make a decision, of course.